### PR TITLE
MM-12244 Adding how to build and deploy a custom server

### PR DESCRIPTION
**Pull request by avasconcelos114**
Original issue: https://github.com/mattermost/mattermost-server/issues/9680

- Building your own server (team edition)
- Editing server files (i18n, email templates)
- Webapp

### CODE REVIEW

**Comment by avasconcelos114** [comment]
This is my first contribution to documentation so I'm not so sure I worded things well enough, I did try to match the overall structure of other pages though.

Let me know if anything needs to be changed, added, or removed :pray:

**Comment by jwilander** [comment]
This is awesome @avasconcelos114! I wonder if it makes to move this from the Contribute section to the Extend section, since this is more about how to extend Mattermost rather than about how to contribute to it. Thoughts?

**Comment by avasconcelos114** [comment]
That actually makes a lot of sense, it probably fits better with people who are looking to expand the functionalities of Mattermost, rather than people who want to contribute to the open source repos.

Shall I go ahead and change that then?

**Comment by jwilander** [comment]
Yeah let's do that 👍

File: site/content/extend/customization/_index.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
+title: "Customizing Mattermost"
+date: "2017-10-29T13:47:55+09:00"
+section: "extend"
+---
+
+Mattermost allows for a variety of customization options and variations, making it possible to create a more adequate experience depending on the needs of each deployment.
```
Comment (amyblais) [inline site/content/extend/customization/_index.md]: "Mattermost allows for a variety of customization options and **modifications**" ...

File: site/content/extend/customization/_index.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+There are a few limitations regarding [how the re-branding of Mattermost](https://www.mattermost.org/trademark-standards-of-use/) must be handled, as well as the fact that changes to the Enterprise Edition's source code isn't supported. However these can be overcome with the utilization of [Plugins](/extend/getting-started).
```
Comment (amyblais) [inline site/content/extend/customization/_index.md]: "However**, these limitations** can be overcome with the utilization" ...

File: site/content/extend/customization/_index.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  
+# Customizable Components
+  
+## Server (Team Edition only)
+The [Mattermost server](https://github.com/mattermost/mattermost-server)'s source code, written in Golang, may be customized to deliver additional functionalities or meet specific security requirements.
```
Comment (amyblais) [inline site/content/extend/customization/_index.md]: ... "customized to deliver additional functionalities or **to** meet specific security requirements. " ...

File: site/content/extend/customization/_index.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  
+# Customizable Components
+  
+## Server (Team Edition only)
+The [Mattermost server](https://github.com/mattermost/mattermost-server)'s source code, written in Golang, may be customized to deliver additional functionalities or meet specific security requirements. 
+  
+It's recommended that you attempt to meet such customizations by leveraging the [Plugin framework](/extend/getting-started) in order to avoid creating any breaking changes, however details on how to build a custom server may be found [here](/contribute/customization/server-build). 
+  
+## Server Files
+Some parts of server-side customizations don't require changes to the source code, view more details on which server files may be customized in [here](/contribute/customization/server-files).
```
Comment (amyblais) [inline site/content/extend/customization/_index.md]: Begin a new sentence with "view more details..."

File: site/content/extend/customization/_index.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  
+It's recommended that you attempt to meet such customizations by leveraging the [Plugin framework](/extend/getting-started) in order to avoid creating any breaking changes, however details on how to build a custom server may be found [here](/contribute/customization/server-build). 
+  
+## Server Files
+Some parts of server-side customizations don't require changes to the source code, view more details on which server files may be customized in [here](/contribute/customization/server-files). 
+
+Note: Modifications to server files can be utilized on both Team Edition and Enterprise Editions.
+
+## Web App
+Mattermost's web application runs on React, and [its codebase](https://github.com/mattermost/mattermost-webapp) has been open-sourced (regardless of which edition your server uses). You can view details on how to customize the web app in [here](/contribute/customization/webapp). Keep in mind, however, that some changes to the webapp can also leverage the [Plugin](/extend/plugins/webapp) framework, which can help reduce the necessity of rebasing your custom client to each Mattermost release.
```
Comment (amyblais) [inline site/content/extend/customization/_index.md]: webapp > web app

File: site/content/extend/customization/server-files.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+```
+cd i18n
+tar -cvf i18n.tar *
+```
+3. Before replacing the files, stop your Mattermost instance if it's currently running.  
+```
+sudo service mattermost stop
+```
+
+4. In your Mattermost deployment, backup and remove the files inside `i18n` that will no longer be in use.
```
Comment (amyblais) [inline site/content/extend/customization/server-files.md]: backup > back up

File: site/content/extend/customization/server-files.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+```
+cd ~/mattermost/i18n
+tar -cvf i18n-yyyy-MM-dd.tar *
+rm -f *.json
+```
+
+4. Transfer your custom `i18n.tar` file to the deployment server and extract it in the `i18n` directory
```
Comment (amyblais) [inline site/content/extend/customization/server-files.md]: Add periods for each step for consistency

File: site/content/extend/customization/server-files.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+date: 2018-05-20T11:35:32-04:00
+weight: 1
+subsection: Customizing Mattermost
+---
+
+In both Mattermost Team and Enterprise editions, you have the freedom to alter how content is displayed by being able to change the contents of localized files and email templates.
+
+Before proceeding with the steps below, make sure you have completed the [mattermost-server](/contribute/server/developer-setup/) and [mattermost-webapp](/contribute/webapp/developer-setup/) setup process.
+## i18n Files
+The `i18n` files define many of the contents seen in email notifications, and responses from the server.
```
Comment (amyblais) [inline site/content/extend/customization/server-files.md]: No comma between "notifications" and "and"

File: site/content/extend/customization/webapp.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+weight: 1
+subsection: Customizing Mattermost
+---
+
+Customizations to the Mattermost Web App can be performed in cases where you need to customize branding, alter localization strings, or fulfill security requirements that are not immediately offered out-of-the-box.  
+  
+## Customization steps 
+With that in mind, customizing and deploying your Mattermost Web App can be done in a few steps:  
+
+1. Fork the [mattermost-webapp](https://github.com/mattermost/mattermost-webapp) repository, and then clone your fork in your local environment.
```
Comment (amyblais) [inline site/content/extend/customization/webapp.md]: Remove comma

File: site/content/extend/customization/webapp.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  
+## Customization steps 
+With that in mind, customizing and deploying your Mattermost Web App can be done in a few steps:  
+
+1. Fork the [mattermost-webapp](https://github.com/mattermost/mattermost-webapp) repository, and then clone your fork in your local environment.  
+```
+git clone https://github.com/<yourgithubusername>/mattermost-webapp
+```
+
+2. Create a separate branch to perform customization tasks in, as it's not recommended to perform them in the `master` branch (more details about that in the next section regarding rebasing).
```
Comment (amyblais) [inline site/content/extend/customization/webapp.md]: "Create a separate branch to perform customization tasks in" - "in" what?

File: site/content/extend/customization/webapp.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+cd ~/mattermost/client && rm -rf *
+```
+
+7. Transfer compressed `dist` files inside Mattermost's `client` directory, and decompress it.  
+```
+tar -xvf dist.tar
+```
+
+## Rebasing to latest version
+Challenges arise when creating a separate custom branch from an active open-source project like `mattermost-webapp`, as the project gets new commits and pull requests on a daily basis, your custom webapp can quickly become outdated.
```
Comment (amyblais) [inline site/content/extend/customization/webapp.md]: Begin a new sentence with "As the project gets..."

File: site/content/extend/customization/webapp.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+```
+
+3. Perform interactive rebase.  
+```
+git checkout custom_branch  
+git rebase -i master
+```  
+Use `git rebase --continue` after resolving any conflicts that arise during rebase process
+
+4. Push new version back to remote (Note: Rebasing requires that you override previous remote version with a `force` push, be sure you've tested that your rebase was successful before completing this last command)
```
Comment (amyblais) [inline site/content/extend/customization/webapp.md]: Begin a new sentence with "Be sure you've..."

**Comment by amyblais** [comment]
@avasconcelos114 Just added suggestions for minor grammar/readability tweaks.

File: site/content/extend/customization/webapp.md (hunk 10 above)
Comment (avasconcelos114) [reply site/content/extend/customization/webapp.md]: Ah, I was under the impression that was a thing people said in English, the intent was to tell people to create a branch for all their customization-related tasks.

I'll then reword this to `Create a separate branch for your customized version`, does that sound good?

**Comment by avasconcelos114** [comment]
Fixed up the issues with grammar/readability you mentioned. You can see those [in this separate commit](https://github.com/mattermost/mattermost-developer-documentation/pull/210/commits/c2efef005a593a685b35511ba5156480aa7a837e).

File: site/content/extend/customization/webapp.md (hunk 10 above)
Comment (amyblais) [reply site/content/extend/customization/webapp.md]: Makes sense now, I read it wrong the first time - works either way then.

**Comment by jwilander** [comment]
Great work on this @avasconcelos114! I think a lot of people are going to find this really helpful
